courseraoauth2client/oauth2.py

```diff
@@ -0,0 +1,504 @@
+            'redirect_uri': self._redirect_uri,
+            'grant_type': 'authorization_code',
+        }
+        return self._request_tokens_from_token_endpoint(form_data)
+
+    def _exchange_refresh_tokens(self):
+        'Exchanges a refresh token for an access token'
+        if self.token_cache is not None and 'refresh' in self.token_cache:
+            # Attempt to use the refresh token to get a new access token.
+            refresh_form = {
+                'grant_type': 'refresh_token',
+                'refresh_token': self.token_cache['refresh'],
+                'client_id': self.client_id,
+                'client_secret': self.client_secret,
+            }
+            try:
+                tokens = self._request_tokens_from_token_endpoint(refresh_form)
+                tokens['refresh'] = self.token_cache['refresh']
+                return tokens
+            except OAuth2Exception:
+                logging.exception(
+                    'Encountered an exception during refresh token flow.')
+        return None
+
+    def _cache_has_good_token(self):
+        'The cache must have a token, and it must expire in > 5 minutes'
+        return (
+            self.token_cache is not None and
+            (time.time() + 5 * 60) < self.token_cache['expires']
+        )
+
+    def build_authorizer(self):
+        if not self._cache_has_good_token():
+            logging.debug('Attempting to use a refresh token.')
+            new_tokens = self._exchange_refresh_tokens()
+            if new_tokens is None:
+                logging.info(
+                    'Attempting to retrieve new tokens from the endpoint. You '
+                    'will be prompted to authorize your'
+                    'app in your web browser.')
+                new_tokens = self._authorize_new_tokens()
+            logging.debug('New tokens: %s', new_tokens)
+            self.token_cache = new_tokens
+        else:
+            logging.debug('Local cache is good.')
+        return CourseraOAuth2Auth(self.token_cache['token'],
+                                  self.token_cache['expires'])
+
+
+def build_oauth2(args=None, cfg=None):
+    if cfg is None:
+        cfg = configuration()
+
+    try:
+        client_id = args.client_id
+    except:
+        client_id = cfg.get('oauth2', 'client_id')
+
+    try:
+        client_secret = args.client_secret
+    except:
+        client_secret = cfg.get('oauth2', 'client_secret')
+
+    try:
+        scopes = args.scopes
+    except:
+        scopes = cfg.get('oauth2', 'scopes')
+
+    try:
+        cache_filename = args.token_cache_file
+    except:
+        cache_filename = cfg.get('oauth2', 'token_cache')
+
+    return CourseraOAuth2(
+        client_id=client_id,
+        client_secret=client_secret,
+        scopes=scopes,
+        token_cache_file=cache_filename
+    )
+
+
+def configuration():
+    'Loads configuration from the file system.'
+    defaults = '''
+[oauth2]
+client_id = NS8qaSX18X_Eu0pyNbLsnA
```

Comment:
Please remove `client_id`, and `client_secret`, as they do not apply to arbitrary apps. Instead, they must be registered via the app console for each individual app.